contracts/src/registry/PermissionedRegistry.sol

```diff
@@ -2,55 +2,295 @@
+            */
+            _copyRoles(getTokenIdResource(ids[i]), from, to, false);
+            _revokeAllRoles(getTokenIdResource(ids[i]), from, false);
+        }
+    }
+
+    /**
+     * @dev Override the base registry _onRolesGranted function to regenerate the token when the roles are granted.
+     */
+    function _onRolesGranted(bytes32 resource, address /*account*/, uint256 /*oldRoles*/, uint256 /*newRoles*/, uint256 /*roleBitmap*/) internal virtual override {
+        uint256 tokenId = getResourceTokenId(resource);
+        // skip just-burn/expired tokens
+        address owner = ownerOf(tokenId);
+        if (owner != address(0)) {
+            _regenerateToken(tokenId, owner);
+        }
+    }
+
+    /**
+     * @dev Override the base registry _onRolesRevoked function to regenerate the token when the roles are revoked.
+     */
+    function _onRolesRevoked(bytes32 resource, address /*account*/, uint256 /*oldRoles*/, uint256 /*newRoles*/, uint256 /*roleBitmap*/) internal virtual override {
+        uint256 tokenId = getResourceTokenId(resource);
+        // skip just-burn/expired tokens
+        address owner = ownerOf(tokenId);
+        if (owner != address(0)) {
+            _regenerateToken(tokenId, owner);
+        }
+    }
+
+    /**
+     * @dev Regenerate a token.
+     */
+    function _regenerateToken(uint256 tokenId, address owner) internal {
+        _burn(owner, tokenId, 1);
+        (address registry, uint64 expires, uint32 tokenIdVersion) = datastore.getSubregistry(tokenId);
+        uint256 newTokenId = _regenerateTokenId(tokenId, registry, expires, tokenIdVersion);
+        _mint(owner, newTokenId, 1, "");
+
+        emit TokenRegenerated(tokenId, newTokenId);
+    }
+
+    /**
+     * @dev Regenerate a token id.
+     * @param tokenId The token id to regenerate.
+     * @param registry The registry to set.
+     * @param expires The expiry date to set.
+     * @param tokenIdVersion The token id version to set.
+     * @return newTokenId The new token id.
+     */
+    function _regenerateTokenId(uint256 tokenId, address registry, uint64 expires, uint32 tokenIdVersion) internal returns (uint256 newTokenId) {
+        tokenIdVersion++;
+        newTokenId = _constructTokenId(tokenId, tokenIdVersion);
+        datastore.setSubregistry(newTokenId, registry, expires, tokenIdVersion);
+    }
+
+    /**
+     * @dev Construct a token id from a canonical/token id and a token id version.
+     * @param id The canonical/token id to construct the token id from.
+     * @param tokenIdVersion The token id version to set.
+     * @return newTokenId The new token id.
+     */
+    function _constructTokenId(uint256 id, uint32 tokenIdVersion) internal pure returns (uint256 newTokenId) {
+        newTokenId = NameUtils.getCanonicalId(id) | tokenIdVersion;
     }
 }
+
+/**
+ * @dev Observer pattern for events on existing tokens.
+ */
+interface TokenObserver {
```

Comment:
This should be in its own file, so that anyone wanting to implement an observer doesn't need to include the entire registry source code.